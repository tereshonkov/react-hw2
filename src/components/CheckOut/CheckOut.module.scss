.wrapper {
    padding: 2% 3%;
    width: 100vw;
}
.header {
    display: flex;
    align-items: center;
    gap: 20px;
    margin-top: 20px;
}
.rectangle {
    display: block;
    width: 6px;
    height: 30px;
    border-radius: 10px;
    background: rgb(138, 51, 253);
}
.wrapperForm {
    display: flex;
}
.form {
    width: 60vw;
    padding: 5% 0;
}
input {
    width: 26vw;
    height: 7vh;
    padding: 5.2%;
    border-radius: 5px;
    border: none;
    background-color: rgb(224, 219, 219);
    box-sizing: border-box;
}
.nameWrapper, .ageWrapper, .adressWrapper {
    display: flex;
    width: 100%;
    gap: 2vw;
    margin-bottom: 50px;
}
.windowFav {
    width: 20vw;
    height: 70vh;
    display: flex;
    flex-direction: column;
    padding: 1%;
    border: 3px solid rgb(227, 223, 223);
    overflow-y: scroll;
}
.line {
    display: block;
    height: 1px;
    width: 100%;
    background-color: rgb(227, 223, 223);
    margin-bottom: 5px;
    margin-top: 5px;
}
.img {
    width: 40px;
    height: 50px;
}
.cart {
    display: flex;
    justify-content: space-between;
    align-items: center;
    gap: 10px;
}
.close {
    border: none;
    cursor: pointer;
}
.btn {
    display: flex;
    align-items: center;
    justify-content: center;
    box-sizing: border-box;
    border: 1px solid rgb(60, 66, 66);
    border-radius: 8px;
    color: rgb(138, 51, 253);
    padding: 16px 20px 16px 20px;
    background-color: transparent;
    max-width: 70%;
    
    &:hover {
        border-radius: 8px;
        background: rgb(138, 51, 253);
        color: rgb(255, 255, 255);
    }
}