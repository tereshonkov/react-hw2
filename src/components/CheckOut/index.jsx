import React, { useReducer } from "react";
import styles from "./CheckOut.module.scss";
import { useDispatch, useSelector } from "react-redux";
import { deleteCart } from "../../redux/slices/productSlice";
import { Formik, Form } from "formik";
import * as Yup from "yup";
import CustomInput from "../CustomInput/СustomInput";
import { deleteAll } from "../../redux/slices/productSlice";

const validationSchema = Yup.object({
  name: Yup.string()
    .min(3, "Імя не може бути таким коротким")
    .max(10, "Імя не може бути таким довгим")
    .required("Це поле обовязкове для заповнення!"),
  lastName: Yup.string()
    .min(2, "Призвіще не може бути таким коротким")
    .max(15, "Призвіще не може бути таким довгим")
    .required("Це поле обовязкове для заповнення!"),
  age: Yup.number()
    .min(16, "Занадто малий вік")
    .max(100, "Вік занадто великий!")
    .required("Це поле обовязкове для заповнення!"),
  phone: Yup.number().min(10, 'Ваш номер не може бути коротший за 10 символів').required('Це поле обовязкове для заповнення!'),
  city: Yup.string()
    .min(3, "Такого міста не існує")
    .max(10, "Такого міста не існує")
    .required("Це поле обовязкове для заповнення!"),
  street: Yup.string()
    .min(3, "Такої вулиці не існує")
    .max(20, "Такої вулиці не існує")
    .required("Це поле обовязкове для заповнення!"),
});

const initialValues = {
  name: localStorage.getItem("name") || "",
  lastName: localStorage.getItem("lastName") || "",
  age: localStorage.getItem("age") || "",
  phone: localStorage.getItem("phone") || "",
  city: localStorage.getItem("city") || "",
  street: localStorage.getItem("street") || "",
};

const saveToLs = (name, value) => {
    localStorage.setItem(name, value.toString());
}
const clearLS = () => {
    localStorage.removeItem("name");
    localStorage.removeItem("lastName");
    localStorage.removeItem("age");
    localStorage.removeItem("phone");
    localStorage.removeItem("city");
    localStorage.removeItem("street");
  };


export default function CheckOut({ firstClick, firstText }) {
  const array = useSelector((state) => state.products.data);
  const cartArray = array.filter((el) => el.quantity > 0);
  let sum = cartArray.reduce((acc, num) => {
    return acc + num.price;
  }, 0);
  sum = Math.round(sum * 100) / 100;
  const dispatch = useDispatch();
  return (
    <div className={styles.wrapper}>
      <div className={styles.headerGroup}>
        <div className={styles.header}>
          <div className={styles.rectangle}></div>
          <h2>Check Out</h2>
        </div>
        <div className={styles.headerSecond}>
          <h2>Billing Details</h2>
        </div>
      </div>
      <div className={styles.wrapperForm}>
        <Formik
          initialValues={initialValues}
          validationSchema={validationSchema}
          onSubmit={(values, {resetForm}) => {
            resetForm();
            clearLS();
            console.log(values);
            console.log(cartArray);
            
            dispatch(deleteAll());
          }}
        >
          <Form className={styles.form} action="">
            <div className={styles.nameWrapper}>
              <CustomInput name="name" label="Name" type="text" placeholder="Enter your name" customChange = {(value, name) => {saveToLs(value, name)}} />
              <CustomInput name="lastName" label="Last Name" type="text" placeholder="Enter your last name" customChange = {(value, name) => {saveToLs(value, name)}}/>
            </div>

            <div className={styles.ageWrapper}>
            <CustomInput name="age" label="Age" type="number" placeholder="Enter your Age" customChange = {(value, name) => {saveToLs(value, name)}}/>
            <CustomInput name="phone" label="Phone Number" type="number" placeholder="Enter your phone number" customChange = {(value, name) => {saveToLs(value, name)}}/>
            </div>

            <div className={styles.adressWrapper}>
            <CustomInput name="city" label="City" type="text" placeholder="Enter your city" customChange = {(value, name) => {saveToLs(value, name)}}/>
            <CustomInput name="street" label="Street" type="text" placeholder="Enter your street" customChange = {(value, name) => {saveToLs(value, name)}}/>
            </div>

            <button className={styles.btn} type="submit">Confirm order</button>
          </Form>
        </Formik>
        <div className={styles.windowFav}>
          <h3>Order Sumary</h3>
          {cartArray.map((el) => (
            <div key={el.id} className={styles.cartWrapper}>
              <div className={styles.line}></div>
              <div className={styles.cart}>
                <img className={styles.img} src={el.img} alt="logo" />
                <p>{el.name}</p>
                <p>{el.price}</p>
                <button
                  onClick={() => {
                    dispatch(deleteCart(el.id));
                  }}
                  className={styles.close}
                >
                  &#10060;
                </button>
              </div>
              <div className={styles.line}></div>
            </div>
          ))}
          <h3>Total price: {sum}</h3>
        </div>
      </div>
    </div>
  );
}
