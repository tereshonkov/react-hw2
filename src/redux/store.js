import { configureStore } from "@reduxjs/toolkit";
import productsSlice from "./slices/productSlice";
import showModalReducer from "./slices/modalSlices";

const store = configureStore({
    reducer: {
        products: productsSlice,
        showModal: showModalReducer,
    }
})

export default store;