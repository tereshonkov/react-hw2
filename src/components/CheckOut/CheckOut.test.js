import { Provider } from "react-redux";
import CheckOut from "./index";
import { render } from "@testing-library/react";
import store from "../../redux/store";

describe("Footer text render", () => {
  test("should render the copyright text", () => {
    const { asFragment } = render(
      <Provider store={store}>
        {" "}
        <CheckOut />{" "}
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
