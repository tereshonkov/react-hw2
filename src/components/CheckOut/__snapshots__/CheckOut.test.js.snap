// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer text render should render the copyright text 1`] = `
<DocumentFragment>
   
  <div
    class="wrapper"
  >
    <div
      class="headerGroup"
    >
      <div
        class="header"
      >
        <div
          class="rectangle"
        />
        <h2>
          Check Out
        </h2>
      </div>
      <div
        class="headerSecond"
      >
        <h2>
          Billing Details
        </h2>
      </div>
    </div>
    <div
      class="wrapperForm"
    >
      <form
        action=""
        class="form"
      >
        <div
          class="nameWrapper"
        >
          <div
            class="inputWrapper"
          >
            <label
              for="name"
            >
              Name
            </label>
            <input
              class=""
              name="name"
              placeholder="Enter your name"
              type="text"
              value=""
            />
          </div>
          <div
            class="inputWrapper"
          >
            <label
              for="lastName"
            >
              Last Name
            </label>
            <input
              class=""
              name="lastName"
              placeholder="Enter your last name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="ageWrapper"
        >
          <div
            class="inputWrapper"
          >
            <label
              for="age"
            >
              Age
            </label>
            <input
              class=""
              name="age"
              placeholder="Enter your Age"
              type="number"
              value=""
            />
          </div>
          <div
            class="inputWrapper"
          >
            <label
              for="phone"
            >
              Phone Number
            </label>
            <input
              class=""
              name="phone"
              placeholder="Enter your phone number"
              type="number"
              value=""
            />
          </div>
        </div>
        <div
          class="adressWrapper"
        >
          <div
            class="inputWrapper"
          >
            <label
              for="city"
            >
              City
            </label>
            <input
              class=""
              name="city"
              placeholder="Enter your city"
              type="text"
              value=""
            />
          </div>
          <div
            class="inputWrapper"
          >
            <label
              for="street"
            >
              Street
            </label>
            <input
              class=""
              name="street"
              placeholder="Enter your street"
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          class="btn"
          type="submit"
        >
          Confirm order
        </button>
      </form>
      <div
        class="windowFav"
      >
        <h3>
          Order Sumary
        </h3>
        <h3>
          Total price: 0
        </h3>
      </div>
    </div>
  </div>
   
</DocumentFragment>
`;
